cmd/bosun/expr/annotate.go

```diff
@@ -0,0 +1,254 @@
+}
+
+func getAndFilterAnnotations(e *State, start, end *time.Time, filter string) (annotate.Annotations, error) {
+	annotations, err := e.AnnotateContext.GetAnnotations(start, end, "", "", "", "", "", "", "")
+	if err != nil {
+		return nil, err
+	}
+	filteredAnnotations := annotate.Annotations{}
+	for _, a := range annotations {
+		if filter == "" {
+			filteredAnnotations = append(filteredAnnotations, a)
+			continue
+		}
+		match, err := boolq.AskExpr(filter, a)
+		if err != nil {
+			return nil, err
+		}
+		if match {
+			filteredAnnotations = append(filteredAnnotations, a)
+		}
+	}
+	sort.Sort(sort.Reverse(annotate.AnnotationsByStartID(filteredAnnotations)))
+	return filteredAnnotations, nil
+}
+
+func AnDurations(e *State, T miniprofiler.Timer, filter, startDuration, endDuration string) (r *Results, err error) {
+	reqStart, reqEnd, err := procDuration(e, startDuration, endDuration)
+	if err != nil {
+		return nil, err
+	}
+	filteredAnnotations, err := getAndFilterAnnotations(e, reqStart, reqEnd, filter)
+	if err != nil {
+		return nil, err
+	}
+	series := make(Series)
+	for i, a := range filteredAnnotations {
+		aStart := a.StartDate.Time
+		aEnd := a.EndDate.Time
+		inBounds := (aStart.After(*reqStart) || aStart == *reqStart) && (aEnd.Before(*reqEnd) || aEnd == *reqEnd)
+		entirelyOutOfBounds := aStart.Before(*reqStart) && aEnd.After(*reqEnd)
+		aDuration := aEnd.Sub(aStart)
+		if inBounds {
+			// time has no meaning here, so we just make the key an index since we don't have an array type
+			series[time.Unix(int64(i), 0).UTC()] = aDuration.Seconds()
+		}
+		if entirelyOutOfBounds {
+			// Duration is equal to that of the full request
+			series[time.Unix(int64(i), 0).UTC()] = reqEnd.Sub(*reqStart).Seconds()
+		}
+		if aDuration == 0 {
+			// This would mean an out of bounds. Should never be here, but if we don't return an error in the case that we do end up here then we might panic on divide by zero later in the code
+			return nil, fmt.Errorf("unexpected annotation with 0 duration outside of request bounds (please file an issue)")
+		}
+		if aStart.Before(*reqStart) {
+			aDurationAfterReqStart := aEnd.Sub(*reqStart)
+			series[time.Unix(int64(i), 0).UTC()] = aDurationAfterReqStart.Seconds()
+			continue
+		}
+		if aEnd.After(*reqEnd) {
+			aDurationBeforeReqEnd := reqEnd.Sub(aStart)
+			series[time.Unix(int64(i), 0).UTC()] = aDurationBeforeReqEnd.Seconds()
+		}
+	}
+	if len(series) == 0 {
+		series[time.Unix(0, 0).UTC()] = math.NaN()
+	}
+	return &Results{
+		Results: []*Result{
+			{Value: series},
+		},
+	}, nil
+}
+
+func AnCounts(e *State, T miniprofiler.Timer, filter, startDuration, endDuration string) (r *Results, err error) {
+	reqStart, reqEnd, err := procDuration(e, startDuration, endDuration)
+	if err != nil {
+		return nil, err
+	}
+	filteredAnnotations, err := getAndFilterAnnotations(e, reqStart, reqEnd, filter)
+	if err != nil {
+		return nil, err
+	}
+	series := make(Series)
+	for i, a := range filteredAnnotations {
```

Comment:
straight up duplication. Makes me sad. could potentially make a helper function that takes a `useFullDuration` flag.

at beginning set `value = (duration or 1)` by flag. And when adding to the series multiply that value by the percentage calculated in the window, and use that for the data point.